src/nvim/buffer.c

```diff
@@ -1899,30 +1902,57 @@ int buflist_getfile(int n, linenr_T lnum, int options, int forceit)
       check_cursor_col();
       curwin->w_cursor.coladd = 0;
       curwin->w_set_curswant = TRUE;
+      win_set_viewport(curwin, wip);
     }
     return OK;
   }
   --RedrawingDisabled;
   return FAIL;
 }
 
+void win_set_viewport(win_T *win, wininfo_T *wip)
+{
+  assert(win != NULL);
+  if (wip == NULL || wip->wi_topline == 0) {
+    return;
+  }
+  set_topline(win, wip->wi_topline);
```

Comment:
`set_topline` is very slow, because it calls `hasFoldingWin .. checkupdate .. foldUpdate . . . syn_get_foldlevel`.